File: Difficulty/Easy/[101] Symmetric Tree/solution.cpp
/*
 * @lc app=leetcode id=101 lang=cpp
 *
 * [101] Symmetric Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
#include <queue>

class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        if (root == nullptr) return true;
        if (root->left == nullptr && root->right == nullptr) return true;
        if (root->left == nullptr || root->right == nullptr) return false;
        queue<TreeNode*> dl, dr;
        TreeNode *l, *r;
        dl.push(root->left);
        dr.push(root->right);
        while (!dl.empty() && !dr.empty()) {
            l = dl.front();
            r = dr.front();
            if (l->val != r->val)
                return false;
            if (l->left != nullptr && r->right != nullptr) {
                dl.push(l->left);
                dr.push(r->right);
            }else if (l->left == nullptr && r->right == nullptr) {}
            else return false;
            if (l->right != nullptr && r->left != nullptr) {
                dl.push(l->right);
                dr.push(r->left);
            }else if (l->right == nullptr && r->left == nullptr) {}
            else return false;
            dl.pop();
            dr.pop();
        }
        return true;
    }
};
// @lc code=end

File: Difficulty/Easy/[101] Symmetric Tree/solution.py
class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        if root is None:
            return True
        if root.left is None and root.right is None:
            return True
        if root.left is None or root.right is None:
            return False
        dl, dr = collections.deque(), collections.deque()
        dl.append(root.left)
        dr.append(root.right)
        while dl and dr:
            l = dl.popleft()
            r = dr.popleft()
            if l.val != r.val:
                return False
            if l.left is not None and r.right is not None:
                dl.append(l.left)
                dr.append(r.right)
            elif l.left is None and r.right is None:
                pass
            else:
                return False
            if l.right is not None and r.left is not None:
                dl.append(l.right)
                dr.append(r.left)
            elif l.right is None and r.left is None:
                pass
            else:
                return False
        return True